### PR TITLE
coprocessor/dag/expr push down builtin method DateDiff

### PR DESCRIPTION
**Pull request by edwardpku**
## What have you changed? (mandatory)

Regarding  #3275 push down builtin method DateDiff

## What are the type of the changes? (mandatory)

Improvement

## How has this PR been tested? (mandatory)

unit test

## Does this PR affect documentation (docs) update? (mandatory)
No

## Does this PR affect tidb-ansible update? (mandatory)

Yes, to make push down works

## Refer to a related PR or issue link (optional)

#3275

### CODE REVIEW

**Comment by ice1000** [comment]
Please sign-off all your commits.

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 1]
```diff
@@ -277,6 +277,20 @@ impl ScalarFunc {
         }
         Ok(Some(result))
     }
+
+    #[inline]
+    pub fn date_diff(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<i64>> {
+        let l_t: Cow<Time> = try_opt!(self.children[0].eval_time(ctx, row));
```
Comment (huachaohuang) [inline src/coprocessor/dag/expr/builtin_time.rs]: Is better to name it `lhs`?

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 2]
```diff
@@ -277,6 +277,20 @@ impl ScalarFunc {
...
+        if l_t.is_zero() {
```
Comment (huachaohuang) [inline src/coprocessor/dag/expr/builtin_time.rs]: I think we should use `invalid_zero()` here. @breeswish

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 3]
```diff
@@ -277,6 +277,20 @@ impl ScalarFunc {
...
+            return handle_incorrect_datetime_error(ctx, l_t).map(|_| None);
+        }
+        let r_t: Cow<Time> = try_opt!(self.children[1].eval_time(ctx, row));
+        if r_t.is_zero() {
+            return handle_incorrect_datetime_error(ctx, r_t).map(|_| None);
+        }
+        let days_diff = l_t.get_time().signed_duration_since(r_t.get_time()).num_days();
```
Comment (huachaohuang) [inline src/coprocessor/dag/expr/builtin_time.rs]: What's the result of `("2018-02-02 00:00:00", "2018-02-01 12:00:00")`?

File: src/coprocessor/dag/expr/builtin_time.rs (hunk 1 above)
Comment (edwardpku) [reply src/coprocessor/dag/expr/builtin_time.rs]: how about lhs_t and rhs_t?

File: src/coprocessor/dag/expr/builtin_time.rs (hunk 2 above)
Comment (edwardpku) [reply src/coprocessor/dag/expr/builtin_time.rs]: Do you mean this function?
https://github.com/tikv/tikv/blob/670a95aadc5ba90cf45b188da639cf7299bf4942/src/coprocessor/codec/mysql/time/mod.rs#L264

I followed convention in the current code, just curious
1) why not extract out a function for DateTime validation. I saw the validation everywhere in the file and can cause inconsistency issue
2) Why zero DateTime is not allowed

File: src/coprocessor/dag/expr/builtin_time.rs (hunk 1 above)
Comment (edwardpku) [reply src/coprocessor/dag/expr/builtin_time.rs]: t could be "type" under some context which is confusing 
may I know is there a naming convention the project is following?

File: src/coprocessor/dag/expr/builtin_time.rs (hunk 3 above)
Comment (edwardpku) [reply src/coprocessor/dag/expr/builtin_time.rs]: Two follow-up questions:
1) In the Go implementation, it's start_date - end_date is it correct?
https://github.com/pingcap/tidb/blob/32b1dbd8d59fbd725ec3f2971da2323cae0bf885/types/mytime.go#L195
2) https://www.w3schools.com/sql/func_sqlserver_datediff.asp
Based on SQL specification it's end_date - start_date and user can select interval(e.g. year, month, day... ) Are we going to support interval?

File: src/coprocessor/dag/expr/builtin_time.rs (hunk 1 above)
Comment (edwardpku) [reply src/coprocessor/dag/expr/builtin_time.rs]: addressed

**Comment by ice1000** [comment]
Sign-off missing

File: src/coprocessor/dag/expr/builtin_time.rs (hunk 3 above)
Comment (huachaohuang) [reply src/coprocessor/dag/expr/builtin_time.rs]: From the [description from MySQL](https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_datediff): "Only the date parts of the values are used in the calculation. "

Here is the output of MySQL:

```
mysql> SELECT DATEDIFF('2018-02-02 00:00:00', '2018-02-01 12:00:00');
+--------------------------------------------------------+
| DATEDIFF('2018-02-02 00:00:00', '2018-02-01 12:00:00') |
+--------------------------------------------------------+
|                                                      1 |
+--------------------------------------------------------+
1 row in set (0.00 sec)
```

So I think there is a misunderstanding here?

File: src/coprocessor/dag/expr/builtin_time.rs (hunk 1 above)
Comment (huachaohuang) [reply src/coprocessor/dag/expr/builtin_time.rs]: May I ask what's the `_t` suffix for?

File: src/coprocessor/dag/expr/builtin_time.rs (hunk 2 above)
Comment (huachaohuang) [reply src/coprocessor/dag/expr/builtin_time.rs]: Maybe @breeswish or @zz-jason can help to answer this.

File: src/coprocessor/dag/expr/builtin_time.rs (hunk 3 above)
Comment (edwardpku) [reply src/coprocessor/dag/expr/builtin_time.rs]: "DATEDIFF() returns expr1 − expr2"  makes sense to me, the naming in KiDB is confusing. Thanks!

**Comment by ice1000** [comment]
Sign-off missing

File: src/coprocessor/dag/expr/builtin_time.rs (hunk 2 above)
Comment (edwardpku) [reply src/coprocessor/dag/expr/builtin_time.rs]: @huachaohuang @breeswish @zz-jason 

DateDiff fixed by always setting time to be zero:
        let lhs_date_time = lhs.get_time().with_hour(0).unwrap()
            .with_minute(0).unwrap().with_second(0).unwrap().with_nanosecond(0).unwrap();

I can submit another CR to refactor DateTime validation. Let me know if you have other comments. 
Gentle reminder, the two questions I raised are still open.

**Comment by edwardpku** [comment]
> Sign-off missing

Hey @ice1000 , I have a stupid question
I'm used to using rebase to update CR
Seems like it doesn't work well with github as I got warnning " Updates were rejected because the remote contains work that you donot have locally."
I didn't get answer from github help: https://help.github.com/articles/about-pull-request-merges/

May I know how you update a PR? Merge? Personally, I don't like merge...

File: src/coprocessor/dag/expr/builtin_time.rs (hunk 2 above)
Comment (XuHuaiyu) [reply src/coprocessor/dag/expr/builtin_time.rs]: Hi, @edwardpku 
> Why zero DateTime is not allowed

MySQL permits you to store dates where the day or month and day are zero in a DATE or DATETIME column. This is useful for applications that need to store birthdates for which you may not know the exact date. In this case, you simply store the date as '2009-00-00' or '2009-01-00'. If you store dates such as these, you should not expect to get correct results for functions such as DATE_SUB() or DATE_ADD() that require complete dates. 

And we check `invalid_zero` here to be compatiable with the behavior of `DATEDIFF` in MySQL.

> why not extract out a function for DateTime validation. I saw the validation everywhere in the file and can cause inconsistency issue

MySQL does not hold a unified validation for Datetime for different situation. For example, different SQL_MODE may cause behavior during calculation for a same query.

**Comment by huachaohuang** [comment]
@edwardpku I think you can try to push with `--force` or `--force-with-lease`.

File: src/coprocessor/dag/expr/builtin_time.rs (hunk 2 above)
Comment (edwardpku) [reply src/coprocessor/dag/expr/builtin_time.rs]: @XuHuaiyu  
Thank you so much for your reply. I made anther revision, please take a look

**Comment by edwardpku** [comment]
> @edwardpku I think you can try to push with `--force` or `--force-with-lease`.

@huachaohuang Thanks for your reply, that's exactly what I did.

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 4]
```diff
@@ -343,6 +344,23 @@ impl ScalarFunc {
     ) -> Result<Option<Cow<'a, Time>>> {
         Ok(Some(Cow::Owned(mysql::time::zero_datetime(ctx.cfg.tz))))
     }
+
+    pub fn date_diff(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<i64>> {
+        let lhs: Cow<Time> = try_opt!(self.children[0].eval_time(ctx, row));
+        if lhs.invalid_zero() {
+            return handle_incorrect_datetime_error(ctx, lhs).map(|_| None);
+        }
+        let lhs_date_time = lhs.get_time().with_hour(0).unwrap()
```
[the remark this answers is not in the document]
Comment (edwardpku) [reply src/coprocessor/dag/expr/builtin_time.rs]: @huachaohuang Thanks! Really helpful suggestion, fixed.

**Comment by huachaohuang** [comment]
The sign-off is missing.

**Comment by huachaohuang** [comment]
@ice1000 PTAL

**Comment by rleungx** [comment]
CI failed. Please fix it.

**Comment by edwardpku** [comment]
@XuHuaiyu @ice1000 @huachaohuang @rleungx 
ci fixed, please take a look

**Comment by rleungx** [comment]
The sign-off is missing. @edwardpku

**Comment by edwardpku** [comment]
@XuHuaiyu @ice1000 @huachaohuang @rleungx 
sign-off added

**Pull request merged**